src/rpc/ngHandlers/NFTBuyOffers.cpp

```diff
@@ -0,0 +1,226 @@
+            JS(destination), toBase58(offer.getAccountID(sfDestination)));
+
+    if (offer.isFieldPresent(sfExpiration))
+        obj.insert_or_assign(JS(expiration), offer.getFieldU32(sfExpiration));
+
+    jv = std::move(obj);
+}
+
+}  // namespace ripple
+
+namespace RPCng {
+
+NFTBuyOffersHandler::Result
+NFTBuyOffersHandler::process(
+    NFTBuyOffersHandler::Input input,
+    boost::asio::yield_context& yield) const
+{
+    auto const tokenID = ripple::uint256{input.nftID.c_str()};
+    auto const range = sharedPtrBackend_->fetchLedgerRange();
+    auto const lgrInfoOrStatus = getLedgerInfoFromHashOrSeq(
+        *sharedPtrBackend_,
+        yield,
+        input.ledgerHash,
+        input.ledgerIndex,
+        range->maxSequence);
+    if (auto const status = std::get_if<Status>(&lgrInfoOrStatus))
+        return Error{*status};
+
+    auto const lgrInfo = std::get<LedgerInfo>(lgrInfoOrStatus);
+    auto const directory = keylet::nft_buys(tokenID);
+
+    // TODO: just check for existence without pulling
+    if (not sharedPtrBackend_->fetchLedgerObject(
+            directory.key, lgrInfo.seq, yield))
+        return Error{Status{RippledError::rpcOBJECT_NOT_FOUND, "notFound"}};
+
+    auto output = NFTBuyOffersHandler::Output{input.nftID};
+    auto offers = std::vector<ripple::SLE>{};
+    auto reserve = input.limit;
+    auto cursor = uint256{};
+    auto startHint = uint64_t{0ul};
+
+    if (input.marker)
+    {
+        cursor = uint256(input.marker->c_str());
+
+        // We have a start point. Use limit - 1 from the result and use the very
+        // last one for the resume.
+        auto const sle =
+            [this, &cursor, &lgrInfo, &yield]() -> std::shared_ptr<SLE const> {
+            auto key = keylet::nftoffer(cursor).key;
+            if (auto const blob = sharedPtrBackend_->fetchLedgerObject(
+                    key, lgrInfo.seq, yield);
+                blob)
+            {
+                return std::make_shared<SLE const>(
+                    SerialIter{blob->data(), blob->size()}, key);
+            }
+            return nullptr;
+        }();
+
+        if (!sle ||
+            sle->getFieldU16(ripple::sfLedgerEntryType) !=
+                ripple::ltNFTOKEN_OFFER ||
+            tokenID != sle->getFieldH256(ripple::sfNFTokenID))
+        {
+            return Error{Status{RippledError::rpcINVALID_PARAMS}};
+        }
+
+        startHint = sle->getFieldU64(ripple::sfNFTokenOfferNode);
+        output.offers.push_back(*sle);
+        offers.reserve(reserve);
+    }
+    else
+    {
+        // We have no start point, limit should be one higher than requested.
+        offers.reserve(++reserve);
+    }
+
+    auto result = traverseOwnedNodes(
+        *sharedPtrBackend_,
+        directory,
+        cursor,
+        startHint,
+        lgrInfo.seq,
+        reserve,
+        {},
+        yield,
+        [&offers](ripple::SLE&& offer) {
+            if (offer.getType() == ripple::ltNFTOKEN_OFFER)
```

Comment:
the directory is only for the nft buy offers, do we still need to check the type here?
If other type is possible, we can not use size check to include marker field (line 151).

Reply:
I have no idea - i just copy the previous implementation and make it testable :)